src/test/java/org/embulk/input/integration/TestFtpFileInputPlugin.java

```diff
@@ -0,0 +1,261 @@
+
+        assertEquals(FTP_PATH_PREFIX + "/sample_01.csv", configDiff.get(String.class, "last_path")); // keep the last_path
+        assertEquals(0, getRecords(config, output).size());
+    }
+
+    @Test
+    public void useAsciiModeTrue()
+            throws Exception
+    {
+        ConfigSource config = this.config.deepCopy().set("ascii_mode", true);
+        ConfigDiff configDiff = runner.transaction(config, new Control());
+
+        assertEquals(FTP_PATH_PREFIX + "/sample_01.csv", configDiff.get(String.class, "last_path"));
+        assertRecords(config, output);
+    }
+
+    @Test
+    public void useSslWithoutSslVerify()
+            throws Exception
+    {
+        ConfigSource config = this.config.deepCopy()
+                .set("port", 990)
+                .set("ssl", true)
+                .set("ssl_verify", false);
+        ConfigDiff configDiff = runner.transaction(config, new Control());
+
+        assertEquals(FTP_PATH_PREFIX + "/sample_01.csv", configDiff.get(String.class, "last_path"));
+        assertRecords(config, output);
+    }
+
+    @Test
+    public void useSslWithCertFile()
+            throws Exception
+    {
+        ConfigSource config = this.config.deepCopy()
+                .set("port", 990)
+                .set("ssl", true)
+                .set("ssl_verify", true)
+                .set("ssl_verify_hostname", "false")
+                .set("ssl_trusted_ca_cert_file", FTP_SSL_TRUSTED_CA_CERT_FILE);
+        ConfigDiff configDiff = runner.transaction(config, new Control());
+
+        assertEquals(FTP_PATH_PREFIX + "/sample_01.csv", configDiff.get(String.class, "last_path"));
+        assertRecords(config, output);
+    }
+
+    @Test
+    public void useSslWithCertData()
+            throws Exception
+    {
+        ConfigSource config = this.config.deepCopy()
+                .set("port", 990)
+                .set("ssl", true)
+                .set("ssl_verify", true)
+                .set("ssl_verify_hostname", "false")
+                .set("ssl_trusted_ca_cert_data", FTP_SSL_TRUSTED_CA_CERT_DATA);
+        ConfigDiff configDiff = runner.transaction(config, new Control());
+
+        assertEquals(FTP_PATH_PREFIX + "/sample_01.csv", configDiff.get(String.class, "last_path"));
+        assertRecords(config, output);
+    }
+
+    private class Control
+            implements InputPlugin.Control
+    {
+        @Override
+        public List<TaskReport> run(TaskSource taskSource, Schema schema, int taskCount)
+        {
+            List<TaskReport> reports = new ArrayList<>();
+            for (int i = 0; i < taskCount; i++) {
+                reports.add(runner.run(taskSource, schema, i, output));
+            }
+            return reports;
+        }
+    }
+
+    private ImmutableMap<String, Object> parserConfig(ImmutableList<Object> schemaConfig)
+    {
+        ImmutableMap.Builder<String, Object> builder = new ImmutableMap.Builder<>();
```

Comment:
Why don't you return ConfigSource instead of Map? I think ConfigSource is more intuitive to represent dynamically-typed map.

```java
return runtime.getExec().newConfigSource()
    .set("type", "csv")
    .set(...)
    ...;
```